### PR TITLE
Storage namespace annotation

### LINKED ISSUE #5490: Storage namespacing
In some cases, such as when loading code from a contract within an LDC proxy, it's possible to mix code that uses the same storage context.

For instance, let's say we have a proxy that uses a single storage slot to store the address it's calling into, and a contract it calls that uses an arbitrary amount of storage.

To maintain coherent execution, the callee must leave the first storage slot untouched.

This creates the possibility for a lot of errors, prevents the future extension of the size of the caller storage context and generally is confusing.

We should instead allow contract storage to be optionally parameterized with a namespace string that acts as a salt when calculating the hash of storage slots, thus preventing collision between storage that uses a different namespace argument.

### PR DESCRIPTION
**Pull request by IGI-111**
## Description

Add a `#[namespace(my_namespace)]` annotation to allow developers to avoid storage slot collisions when loading contract code using LDC.

Fixes #5490

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Review by xunilrj — APPROVED** [review]

File: sway-core/src/ir_generation/function.rs [hunk 1]
```diff
@@ -2970,7 +2972,7 @@ impl<'eng> FnCompiler<'eng> {
 
         // Do the actual work. This is a recursive function because we want to drill down
         // to load each primitive type in the storage field in its own storage slot.
-        self.compile_storage_read(context, ix, &field_idcs, &base_type, span_md_idx)
+        self.compile_storage_read(context, ns, ix, &field_idcs, &base_type, span_md_idx)
```
Comment (ironcev) [inline sway-core/src/ir_generation/function.rs:2975]: Nitpick. Just a slight inconsistency, `compile_storage_access` accepts `fields`, `ix`, and `ns` in that order, whereas other methods go for `ns`, `ix`, `indices` (gotten from `fields`) essentially other way around (and more logical to me, to start with the namespace).

File: sway-core/src/semantic_analysis/ast_node/declaration/storage.rs [hunk 2]
```diff
@@ -58,6 +59,7 @@ impl ty::TyStorageField {
         md_mgr: &mut MetadataManager,
         module: Module,
         ix: &StateIndex,
+        namespace: Option<&str>,
```
Comment (ironcev) [inline sway-core/src/semantic_analysis/ast_node/declaration/storage.rs:62]: While I am at nit picking :smile: here as well, `namespace` first (or `ns` to be fully consistent?).

File: test/src/sdk-harness/test_projects/storage_namespace/.gitignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+out
```
Comment (ironcev) [inline test/src/sdk-harness/test_projects/storage_namespace/.gitignore:1]: Local .gitignore file not needed.

**Review by ironcev — APPROVED** [review]
Few nit picks and one redundant `.gitignore` file worth deleting, but otherwise LGTM very straightforward implementation.

**Review by ironcev — APPROVED** [review]

**Review by JoshuaBatty — APPROVED** [review]

**Pull request merged**